infix/lexer.go

```diff
@@ -0,0 +1,225 @@
+package infix
+
+import (
+	"fmt"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+)
+
+type (
+	lexer struct {
+		input  string // expression being lex'ed
+		start  int    // start position of token
+		pos    int    // pos in the input
+		width  int    // width of last rune
+		tokens chan Tokval
+	}
+
+	stateFn func(*lexer) stateFn
+
+	// Tokval is the token type + value
+	Tokval struct {
+		T Token
+		V string
+	}
+)
+
+// eof mimics the EOF but for the input string.
+// The name isn't eoi to avoid confusion.
+const eof = -1
+
+func (t Tokval) String() string {
+	return fmt.Sprintf("Token(%s, %s)", t.T, t.V)
+}
+
+// Lex creates a concurrent lexer and returns a
+// channel of tokens processed from input.
+func Lex(input string) chan Tokval {
+	l := &lexer{
+		input:  input,
+		tokens: make(chan Tokval),
+	}
+
+	go l.run()
+	return l.tokens
+}
+
+// run the state machine
+func (l *lexer) run() {
+	for state := lexStart; state != nil; {
+		state = state(l)
+	}
+	close(l.tokens)
+}
+
+// emit a token.
+func (l *lexer) emit(tok Token) {
+	l.tokens <- Tokval{
+		T: tok,
+		V: l.input[l.start:l.pos],
+	}
+	l.start = l.pos
+}
+
+// errorf emits an illegal token. This token carries
+// the lexer error.
+func (l *lexer) errorf(msg string, args ...interface{}) stateFn {
+	l.tokens <- Tokval{
+		T: Illegal,
+		V: fmt.Sprintf(msg, args...),
+	}
+	return nil
+}
+
+// peek looks for the next rune from the input
+// but do not increases the cursor.
+// It returns eof when reaches the end of input.
+func (l *lexer) peek() rune {
+	r := l.next()
+	if r == eof {
+		return eof
+	}
+	l.backup()
+	return r
+}
+
+// next returns the next rune from input or
+// eof if reached end of input.
+func (l *lexer) next() (r rune) {
+	if l.pos >= len(l.input) {
+		// zeroes the width to simplify handling at
+		// end of input. Some functions like acceptRun*
+		// do a explicit rewind/backup of last processed
+		// character, but in case of eof we dont want
+		// that.
+		l.width = 0
+		return eof
+	}
+	r, l.width = utf8.DecodeRuneInString(l.input[l.pos:])
+	l.pos += l.width
+	return r
+}
+
+// backup rollback one rune.
+func (l *lexer) backup() {
+	l.pos -= l.width
+}
+
+// ignore bytes processed since last token.
+func (l *lexer) ignore() {
+	l.start = l.pos
+}
+
+// accept checks if next is in the valid string.
+func (l *lexer) accept(valid string) bool {
+	return strings.IndexRune(valid, l.peek()) != -1
+}
+
+// acceptRun consumes next runes if valid.
+func (l *lexer) acceptRun(valid string) {
```

Comment:
because it accepts a run of runes =P 
A run of runes in the sense of a series of runes. I can rename if you like.

Reply:
Never heard of the term "run of runes", perhaps that is why I did not understood it =)

Reply:
got this naming from Rob Pike presentation on writing lexers. The name did makes sense at the time probably because I often see it being used in math, as in _a run of consecutive numbers_.

Reply:
Hm, but a **run** itself is a consecutive sequence of numbers. This is your [definition](http://www.maths.surrey.ac.uk/hosted-sites/R.Knott/runsums/index.html):

>  A run of numbers is a sequence of consecutive whole numbers i.e. with no gaps in the sequence:  such as [2, 3, 4], but not [5, 6, 8, 9] (because 7 is missing) either [2, 3, 3] (repeated 3)

I don't see the mathematical `run` making sense with runes, except if run was redefined by someone like Rob Pike does and now it have a more broad definition.

Wouldn't better just use a more generic concept and common in programming? Like sequence, I don't know. `matchSeq`or something like that.